viewing_party/party.py

```diff
@@ -1,23 +1,206 @@
+
+    return user_data
+
+        
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    sum_rating = 0.0
+    for dict in user_data["watched"]:
+        sum_rating += dict["rating"]
+    
+    try:
+        avg_rating = sum_rating / len(user_data["watched"])
+    except ZeroDivisionError:
+        return 0.0
 
+    return avg_rating
+
+
+def get_most_watched_genre(user_data):
+    genre_dict = {}
+
+    if len(user_data["watched"]) == 0:
+        return None
+    
+    for dict in user_data["watched"]:
+        if not dict["genre"] in genre_dict:
+            genre_dict[dict["genre"]] = 1
+
+        elif dict["genre"] in genre_dict:
+            genre_dict[dict["genre"]] += 1
+    
+    most_watched = ["", 0]
+
+    for key,value in genre_dict.items():
+        if value > most_watched[1]:
+            most_watched[0] = key
+            most_watched[1] = value
+
+    return most_watched[0]
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
+def get_unique_watched(user_data):
+    user_list = []
+    friends_list = []
+    final_list = []
+
+    for dict in user_data["watched"]:
+        user_list.append(dict["title"])
         
+    for movie_dict in user_data["friends"]:
+        for i in range(len(user_data["friends"])):
+            for movie_dict in user_data["friends"][i]["watched"]:
+                friends_list.append(movie_dict["title"])
+
+    user_set = set(user_list)
+    friends_set = set(friends_list)
+
+    unique_movies = user_set - friends_set
+    unique_movies_list = list(unique_movies)
+
+    for movie_dict in user_data["watched"]:
+        for unique_movie in unique_movies_list:
+            if unique_movie in movie_dict["title"]:
+                final_list.append(movie_dict)
+            
+    return final_list
+
+
+    
+def get_friends_unique_watched(user_data):
+    user_list = []
+    friends_list = []
+    final_list = []
+
+    for dict in user_data["watched"]:
+        user_list.append(dict["title"])
+        
+    for movie_dict in user_data["friends"]:
+        for i in range(len(user_data["friends"])):
+            for movie_dict in user_data["friends"][i]["watched"]:
+                print(movie_dict["title"])
+                friends_list.append(movie_dict["title"])
+
+    user_set = set(user_list)
+    friends_set = set(friends_list)
+
+    unique_movies = friends_set - user_set
+    unique_movies_list = list(unique_movies)
+
+    for movie_dict in user_data["friends"]:
+        for i in range(len(user_data["friends"])):
+            for movie_dict_2 in user_data["friends"][i]["watched"]:
+                for unique_movie in unique_movies_list:
+                    if unique_movie in movie_dict_2["title"] and not movie_dict_2 in final_list:
+                        final_list.append(movie_dict_2)

```

Comment:
Same here

```python
    for friend in user_data["friends"]:
        for movie_dict in friend["watched"]:
            for unique_movie in unique_movies_list:
                if unique_movie in movie_dict["title"] and not movie_dict in final_list:
                    final_list.append(movie_dict)
```

Reply:
In your `if` statement, the `in` operator on a `list` hides O(n) time complexity! How might you use sets to reduce the time complexity here?